### DISCUSSION TITLE
ASP.NET Core - Why async await is useful

### DISCUSSION THREAD

**Discussion by giscus** [linkdotnet/Blog.Discussions#19] [General]
# ASP.NET Core - Why async await is useful

Did you ever wonder why you "should" use async and await in your ASP.NET Core applications? Most probable you heard something about performance. And there is some truth to it, but not in the way you might think.
So let's discuss this with smaller examples.

https://steven-giesel.com/blogPost/b925c8c1-03ba-4eb6-9b81-f09df56e0158

**Comment by FoxDawg** [comment]
Good call! You might want to check out my [article](https://dateo-software.de/blog/best-practices-async-await/) on async-await as well.
Best, 
Dennis

**Reply by linkdotnet** [reply]
Hey @FoxDawg  thanks for sharing your article.
Nice write-up. I am not really friend of number 4 (Don’t await too much, return the task object instead), because your method `MakeCoffeeAsync` will not show up in a stack trace (if the caller awaits that call), which makes debugging harder.
If you have a high-throughput endpoint, where every ms matters, maybe than one could do this (but even then async calls are normally to db's, meaning they are at least a few magnitude slower anyway).

**Comment by Ma7moudZakaria** [comment]
I have one question,  

Imagine you have a server which only has 1 available worker thread in the pool. Also two requests are coming in shortly one after another, 
Okay, here we will use asynchronous, right so here is my question the thread will excute this method await repository.GetByIdAsync(...) of the first request then we have the second request that will call this method for example await repository.GetAllAsync(...) here in this time that the thread excute this line the result of first request is back and the thread excute the method of the second request. 

what will happened in this case

**Reply by linkdotnet** [reply]
Very good question. As a small precondition, both methods don't need a thread involved and are I/O bound.
So you have one thread and one thread only. You called `GetByIdAsync` first, which is currently handled by your database, therefore the thread is free and can take up the request for `GetAllAsync`.

Now your db returns the results to your operating system (the `GetByIdAsync`), but there is still no available thread, so it stay's in a queue until there is one free. Luckily we hit `GetAllAsync` where .NET decides to free up that one poor thread to the thread-pool. This then can be immediately picked up by our first request (`GetByIdAsync`) to do the rest of the logic. Once this is done, the thread can handle the other request (`GetAllAsync`).

The gist is, that you can handle both request, but often times you have to wait for each other.

Does that answer your question @Ma7moudZakaria ?

**Comment by dmitry-pavlov** [comment]
## A few notes: 
- no `CancellationToken` passed to async action. Wut?! :open_mouth:
- `.ConfigureAwait(...)` makes no sense as there is no synchronization context in `ASP.NET Core` :angry:

> "Never underestimate the power of getting familiar with the official documentation" (c) Dmitry Pavlov aka the [Coding Machine](https://Coding-Machine.NET)

**Reply by linkdotnet** [reply]
I removed the `ConfigureAwait`. Yes it has literally zero impact, because there is no SynchronizationContext around. I am mainly a library maintainer and that is just something you "do" when you write library code :D 

To your first point: The code was not really meant to show someone, how to write a perfect controller, it was more about the difference of an async code-block vs a sync code-block when one or multiple requests arrive.

I admit that this even speaks more in favour of the async version and I might append that part.
